add case test_volume_metrics

chriscchien opened the pull request:
Add test case `test_volume_metrics`

ref: [4029](https://github.com/longhorn/longhorn/issues/4029)

Each metirc has validated values in  label `node`, `pvc` and `volume`

And for the metric value validation please see below table
| metric  | description |  
|---|---|
| longhorn_volume_actual_size_bytes | assert `volume.controllers[0].actualSize ` | 
| longhorn_volume_capacity_bytes  | assert `volume.size` | 
| longhorn_volume_state |  some items not test because timing issue <br> &#10004; attached <br /> <br> &#10004; detached <br /> <br> &#10007; creating <br /> <br> &#10007; attaching <br /> <br> &#10007; etaching <br /> <br> &#10007; deleting </br>| 
| longhorn_volume_robustness| the item not test is because pytest did not support node down scenario<br> &#10004; healthy </br> <br> &#10004; degraded </br> <br> &#10004; faulted </br> <br> &#10007; unknown </br> |
| longhorn_volume_read_throughput  | assert value >= `0.0` | 
| longhorn_volume_write_throughput | assert value >= `0.0` | 
| longhorn_volume_read_iops  | assert value >= `0.0` | 
| longhorn_volume_write_iops  | assert value >= `0.0` | 
| longhorn_volume_read_latency  | assert value >= `0.0` | 
| longhorn_volume_write_latency  | assert value >= `0.0` | 


test [result](https://ci.longhorn.io/job/private/job/longhorn-tests-regression/4930/testReport/tests/test_metric/)

khushboo-rancher reviewed (commented):
In general, looks good. Few minor comments and a doubt as below.

Can you add 'unknown` check for `volume.status.robustness` for a detached volume?
~Can we try adding checks for `attaching, detaching, creating and deleting` in a loop. Like trying to catch the transitional state three times.
Also, we can organize the test block into sections (for state, robustness etc) to increase readability. Let me know if you have doubt.~
Edit: Removed the request to check the transient state as @PhanLe1010 suggested. :)

chriscchien commented:
I will update this PR with `pvc_namespace` check regarding the [ticket](https://github.com/longhorn/longhorn/issues/7088)

chriscchien commented on manager/integration/tests/test_metric.py:
```diff
@@ -0,0 +1,129 @@
+from common import wait_for_volume_faulted, delete_replica_processes
+from common import Mi
+
+longhorn_volume_state = {
+    "creating": 1.0,
+    "attached": 2.0,
+    "detached": 3.0,
+    "attaching": 4.0,
+    "detaching": 5.0,
+    "deleting": 6.0,
+    }
+
+longhorn_volume_robustness = {
+    "unknown": 0.0,
+    "healthy": 1.0,
+    "degraded": 2.0,
+    "faulted": 3.0,
+}
```
Hi @khushboo-rancher ,

The value is from official [website](https://longhorn.io/docs/1.5.3/monitoring/metrics/#volume) and were used to verify metric return value. 
![image](https://github.com/longhorn/longhorn-tests/assets/73337386/9e5bbf3a-cc92-4d7f-b345-24ac65d111c1)

chriscchien commented:
> In general, looks good. Few minor comments and a doubt as below.
> 
> ~Can we try adding checks for `attaching, detaching, creating and deleting` in a loop. Like trying to catch the transitional state three times.~ Can we try to organize the test block into sections (for state, robustness etc) to increase readability. Let me know if you have doubt. Can you add `unknown` check for `volume.status.robustness` for a detached volume?
> 
> Edit: Removed the request to check the transient state as @PhanLe1010 suggested. :)

Updated and here is 20 times test [result](https://ci.longhorn.io/job/private/job/longhorn-tests-regression/5275/testReport/tests/test_metric/)

yangchiu commented:
@chriscchien Is it possible to add scenario for volume's PVC in other non-default namespaces like https://github.com/longhorn/longhorn/issues/7077#issuecomment-1812150822? Many Thanks.

chriscchien commented:
> @chriscchien Is it possible to add scenario for volume's PVC in other non-default namespaces like [longhorn/longhorn#7077 (comment)](https://github.com/longhorn/longhorn/issues/7077#issuecomment-1812150822)? Many Thanks.

Updated and test [result](https://ci.longhorn.io/job/private/job/longhorn-tests-regression/5312/testReport/tests/test_metric/)

yangchiu commented:
@chriscchien Please request a review to make this PR approved.

yangchiu reviewed (approved):
LGTM

nitendra-suse commented on manager/integration/tests/test_metric.py:
```diff
@@ -0,0 +1,159 @@
+from common import wait_for_volume_detached_unknown
+from common import LONGHORN_NAMESPACE
+
+longhorn_volume_state = {
```
Can we instead generate these dynamically and use integers instead of floats?
i.e

`longhorn_volume_state = {vol_state:index for index,vol_state in enumerate(["creating","attached","detached","attaching","detaching","deleting"],1)}`

chriscchien replied:
Updated

c3y1huang replied:
From my perspective, creating this dynamically seems to be unnecessarily complicating the code. @nitendra-suse , could you provide more insight into why dynamic generation is beneficial here?

nitendra-suse replied:
@c3y1huang I feel dynamical version is idiomatic python, But I agree your implementation is more straightforward, we can keep it.

c3y1huang commented on manager/integration/tests/test_metric.py:
```diff
@@ -0,0 +1,159 @@
+def get_metrics(core_api): # NOQA
+    lht_hostId = get_self_host_id()
+
+    pods = core_api.list_namespaced_pod(namespace="longhorn-system",
```
Suggest using `LONGHORN_NAMESPACE` in common.py.

chriscchien replied:
Updated

c3y1huang commented on manager/integration/tests/test_metric.py:
```diff
@@ -0,0 +1,159 @@
+
+    pods = core_api.list_namespaced_pod(namespace="longhorn-system",
+                                        label_selector="app=longhorn-manager")
+    for pod in pods.items:
+        if pod.spec.node_name == lht_hostId:
+            manager_ip = pod.status.pod_ip
+            break
```
Probably can reuse `get_mgr_ips` in common.py.

chriscchien replied:
I want to get IP address of the `longhorn-manager` pod that is running on the same node as the `longhorn-test` pod (volume attached on that node too) but can not figure a way to leverage `get_mgr_ips`, could you give me more guidance? thank you .

c3y1huang replied:
Got it, I overlooked the `longhorn-test` pod condition check. It's clear now.

c3y1huang commented on manager/integration/tests/test_metric.py:
```diff
@@ -0,0 +1,159 @@
+                item = sample
+                break
+
+    assert found is True
```
Assert `item` is not `None` might achieve the same thing.

chriscchien replied:
Updated

c3y1huang commented on manager/integration/tests/test_metric.py:
```diff
@@ -0,0 +1,159 @@
+import pytest
+import requests
+from prometheus_client.parser import text_string_to_metric_families
+from common import client, core_api, volume_name  # NOQA
+from common import get_self_host_id
+from common import create_pv_for_volume, create_pvc_for_volume
+from common import create_and_check_volume, wait_for_volume_healthy
+from common import write_volume_random_data
+from common import wait_for_volume_detached, wait_for_volume_degraded
+from common import wait_for_volume_faulted, crash_replica_processes
+from common import Mi
+from common import wait_for_volume_detached_unknown
+from common import LONGHORN_NAMESPACE
```
nit: consider reorganize the imports.

chriscchien commented on manager/integration/tests/test_metric.py:
```diff
@@ -0,0 +1,155 @@
+from common import wait_for_volume_detached_unknown
+from common import Mi, LONGHORN_NAMESPACE
+
+longhorn_volume_state = {
+    "creating": 1.0,
+    "attached": 2.0,
+    "detached": 3.0,
+    "attaching": 4.0,
+    "detaching": 5.0,
+    "deleting": 6.0,
+    }
+
+longhorn_volume_robustness = {
+    "unknown": 0.0,
+    "healthy": 1.0,
+    "degraded": 2.0,
+    "faulted": 3.0,
+}
```
Updated and here is the test [result](https://ci.longhorn.io/job/private/job/longhorn-tests-regression/5334/testReport/tests/test_metric/).

chriscchien commented:
@mergify backport v1.5.x v1.4.x
